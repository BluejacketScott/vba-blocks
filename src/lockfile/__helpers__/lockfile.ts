export function normalizeLockfile(lockfile: string): string {
	return lockfile.replace(/vba\-blocks v\d.\d.\d/g, "vba-blocks v#.#.#");
}
