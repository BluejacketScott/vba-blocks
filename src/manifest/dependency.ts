import { manifestOk } from '../errors';
import has from '../utils/has';
import { isString } from '../utils/is';
import { extname, join, trailing } from '../utils/path';
import { Version } from './version';

/*
  # Dependency

  Three types of dependencies: registry, path, and git

  - registry: version | { version, registry? }

    Registry (default = "vba-blocks"):
    index + packages are loading in config and used for resolve and fetching

  - path: path
  - git: { git, branch?, tag?, rev? }

    Where git is path to git remote and revision is pulled from
    master, branch, tag, or rev
*/

export interface DependencyDetails {
  name: string;
  version?: Version;
}

export interface RegistryDependency extends DependencyDetails {
  registry: string;
  version: Version;
}

export interface PathDependency extends DependencyDetails {
  path: string;
}

export interface GitDependency extends DependencyDetails {
  git: string;
  tag?: string;
  branch?: string;
  rev?: string;
}

export type Dependency = RegistryDependency | PathDependency | GitDependency;

const EXAMPLE = `Example vba-block.toml:

  [dependencies]
  a = "^1.0.0"
  b = { version = "^0.1.0" }
  c = { path = "packages/c" }
  d = { git = "https://github.com/author/d" }
  e = { git = "https://github.com/author/e", branch = "next" }
  f = { git = "https://github.com/author/f", tag = "v1.0.0" }
  g = { git = "https://github.com/author/g", rev = "a1b2c3d4" }

  [dependencies.h]
  version = "^2.0.0"`;

export function parseDependencies(value: any, dir: string): Dependency[] {
  return Object.entries(value).map(([name, value]) => parseDependency(name, value, dir));
}

export function parseDependency(name: string, value: Version | any, dir: string): Dependency {
  if (isString(value)) value = { version: value };

  let {
    registry = 'vba-blocks',
    version,
    path,
    git,
    tag,
    branch = 'master',
    rev
  }: {
    registry?: string;
    version?: string;
    path?: string;
    git?: string;
    tag?: string;
    branch?: string;
    rev?: string;
  } = value;

  manifestOk(
    version || path || git,
    `Invalid dependency "${name}", no version, path, or git specified. \n\n${EXAMPLE}`
  );

  if (version) {
    return { name, registry, version };
  } else if (path) {
    return { name, path: trailing(join(dir, path)) };
  } else {
    git = ensureGitUrl(git!);

    if (rev) return { name, git, rev };
    else if (tag) return { name, git, tag };
    else return { name, git, branch };
  }
}

export function isRegistryDependency(dependency: Dependency): dependency is RegistryDependency {
  return has(dependency, 'registry');
}

export function isPathDependency(dependency: Dependency): dependency is PathDependency {
  return has(dependency, 'path');
}

export function isGitDependency(dependency: Dependency): dependency is GitDependency {
  return has(dependency, 'git');
}

export function formatDependencies(dependencies: Dependency[]): object {
  const value: { [name: string]: any } = {};
  dependencies.forEach(dependency => {
    if (isRegistryDependency(dependency)) {
      const { name, registry, version } = dependency;
      value[name] = registry !== 'vba-blocks' ? { version, registry } : version;
    } else if (isPathDependency(dependency)) {
      const { name, path } = dependency;
      value[name] = path;
    } else {
      const { name, git, tag, branch, rev } = dependency;
      value[name] = { name, git };
      if (tag) value[name].tag = tag;
      if (branch) value[name].branch = branch;
      if (rev) value[name].rev = rev;
    }
  });

  return value;
}

function ensureGitUrl(value: string): string {
  if (extname(value!) === '.git') return value;

  return `${value}.git`;
}
