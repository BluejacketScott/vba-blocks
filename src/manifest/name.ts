import dedent from '@timhall/dedent/macro';
import { CliError, ErrorCode } from '../errors';

/*
  # Project/package name

  Name requirements:
  - Must not be empty
  - Allowed characters: A-Z a-z 0-9 - _ . /
  - Must start with an alphanumeric
  - A single / is allowed for scoped packages
  - A maximum length of 255 for the scope and 255 for the package name

  Note:
  - Capitals are allowed, but compared and stored as lowercase
*/

const ONLY_VALID_CHARACTERS = /^[A-Za-z0-9-_./]*$/;
const STARTS_WITH_ALPHANUMERIC = /^[A-Za-z0-9]/;
const MAX_LENGTH = 255;

export interface ManifestName {
  scope: string | undefined;
  name: string;
  parts: string[];
}

export function parseName(name: string): ManifestName {
  assertValidName(name);

  // Manifest name is case-insensitve,
  // but parsed name isn't for comparison, uniqueness, and file naming
  name = name.toLowerCase();

  let scope: string | undefined;
  let pkg: string;
  if (name.includes('/')) {
    [scope, pkg] = name.split('/');
  } else {
    pkg = name;
  }

  return { scope, name: pkg, parts: scope ? [scope, pkg] : [pkg] };
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

export function validateName(name: string): ValidationResult {
  const errors = [];

  if (!name || typeof name !== 'string' || !name.length) {
    errors.push('name must not be empty');
  }
  if (!ONLY_VALID_CHARACTERS.test(name)) {
    errors.push('name must only include the following characters: A-Z a-z 0-9 - _ . /');
  }
  if (!STARTS_WITH_ALPHANUMERIC.test(name)) {
    errors.push('name must only start with the following characters: A-Z a-z 0-9');
  }
  if (countMatches(name, /\//g) > 1) {
    errors.push('name must only include a single / for scoped packages');
  }

  let scope: string | undefined;
  let pkg: string;
  if (name.includes('/')) {
    [scope, pkg] = name.split('/');
  } else {
    pkg = name;
  }

  if (scope && scope.length > MAX_LENGTH) {
    errors.push(`scope must be less than ${MAX_LENGTH} characters`);
  }
  if (pkg.length > MAX_LENGTH) {
    errors.push(`name must be less than ${MAX_LENGTH} characters`);
  }

  return { valid: errors.length === 0, errors };
}

export function assertValidName(name: string) {
  const { valid, errors } = validateName(name);
  if (valid) return;

  const error_messages = errors.map(error => `- ${error}`).join('\n');

  throw new CliError(
    ErrorCode.ManifestNameInvalid,
    dedent`
      Invalid project/package name "${name}"

      Errors:
      ${error_messages}
    `
  );
}

function countMatches(value: string, search: RegExp): number {
  return (value.match(search) || []).length;
}
