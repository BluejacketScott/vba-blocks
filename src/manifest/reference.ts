import { manifestOk } from '../errors';

/*
  # Reference

  To add reference, guid, major version, and minor version are required

  reference: { guid, version }

  Where version is "MAJOR.MINOR"
*/

export interface ReferenceDetails {
  dependency?: string;
}

export interface Reference {
  name: string;
  guid: string;
  major: number;
  minor: number;
}

const VERSION_REGEX = /^(\d+)\.(\d+)$/;
const GUID_REGEX = /\{.{8}-.{4}-.{4}-.{4}-.{12}\}/;

const isVersion = (value: string) => VERSION_REGEX.test(value);
const isGuid = (value: string) => GUID_REGEX.test(value);

const toInt = (value: string) => parseInt(value, 10);
const getMajorMinor = (version: string) => {
  const [major, minor] = version.split('.', 2).map(toInt);

  return { major, minor };
};

const EXAMPLE = `Example vba-block.toml:

  [references.Scripting]
  version = "1.0"
  guid = "{420B2830-E718-11CF-893D-00A0C9054228}"`;

export function parseReferences(value: any): Reference[] {
  return Object.entries(value).map(([name, value]) => parseReference(name, value));
}

export function parseReference(name: string, value: any): Reference {
  const { version, guid } = value;

  manifestOk(
    isVersion(version),
    `Reference "${name}" has an invalid version "${version}". \n\n${EXAMPLE}.`
  );
  manifestOk(isGuid(guid), `Reference "${name}" has an invalid guid "${guid}". \n\n${EXAMPLE}'`);

  const { major, minor } = getMajorMinor(version);

  return { name, guid, major, minor };
}
