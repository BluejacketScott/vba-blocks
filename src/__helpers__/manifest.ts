import { dir as FIXTURES } from '../../tests/__fixtures__';
import { Manifest } from '../manifest';
import { isPathDependency } from '../manifest/dependency';
import { relative } from '../utils/path';

export function normalizeManifest(manifest: Manifest, relativeTo: string = FIXTURES): Manifest {
  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }
  for (const src of manifest.devSrc) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.devDependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  if (manifest.target) {
    manifest.target.path = normalizePath(manifest.target.path, relativeTo);
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return relative(relativeTo, path);
}
