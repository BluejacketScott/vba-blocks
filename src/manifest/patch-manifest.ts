import dedent from 'dedent';
import { relative } from '../utils/path';
import { Manifest, Source, Dependency, Target } from './';
import { isRegistryDependency, isPathDependency } from './dependency';
import {
  patchApplyChanges,
  patchAddSrc,
  patchRemoveSrc,
  patchAddDependency,
  patchRemoveDependency
} from '../messages';
import env from '../env';

export function applyChanges(changes: string[]) {
  if (!changes.length) return;

  env.reporter.log(`${patchApplyChanges()}\n\n${changes.join('\n\n')}`);
}

export function addSource(manifest: Manifest, source: Source): string {
  const relative_path = relative(manifest.dir, source.path);
  const relative_binary_path = source.binary && relative(manifest.dir, source.binary);

  const details = relative_binary_path
    ? `{ path = "${relative_path}", binary = "${relative_binary_path}" }`
    : `"${relative_path}"`;

  return dedent`
    ${patchAddSrc()}
    ${source.name} = ${details}`;
}

export function removeSource(_: Manifest, name: string): string {
  return patchRemoveSrc(name);
}

export function addDependency(manifest: Manifest, dependency: Dependency): string {
  let details;
  if (isRegistryDependency(dependency)) {
    const { version, registry } = dependency;
    details =
      registry !== 'vba-blocks'
        ? `{ version = "${version}", registry = "${registry}" }`
        : `"${version}"`;
  } else if (isPathDependency(dependency)) {
    const relative_path = relative(manifest.dir, dependency.path);
    details = `{ path = "${relative_path}" }`;
  } else {
    const { git, tag, branch, rev } = dependency;
    if (rev) {
      details = `{ git = "${git}", rev = "${rev}" }`;
    } else if (tag) {
      details = `{ git = "${git}", tag = "${tag}" }`;
    } else if (branch !== 'master') {
      details = `{ git = "${git}", branch = "${branch}" }`;
    } else {
      details = `{ git = "${git}" }`;
    }
  }

  return dedent`
    ${patchAddDependency()}
    ${dependency.name} = ${details}`;
}

export function removeDependency(_: Manifest, name: string): string {
  return patchRemoveDependency(name);
}

export function addTarget(manifest: Manifest, target: Target): string {
  const relative_path = relative(manifest.dir, target.path);

  let details;
  if (target.name !== manifest.name && relative_path !== 'target') {
    details = target.type;
  } else {
    details = `{ type = "${target.type}"`;
    if (target.name !== manifest.name) details += `, name = "${target.name}"`;
    if (relative_path !== 'target') details += `, path = "${relative_path}"`;
    details += ' }';
  }

  return dedent`
    Add the following to the [${manifest.type}] section:
    target = ${details}`;
}
