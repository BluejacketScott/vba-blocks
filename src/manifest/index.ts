import { join, normalize, relative } from '../utils/path';
import { parse as parseToml, convert as convertToToml } from '../utils/toml';
import { pathExists, readFile, writeFile } from '../utils/fs';
import { manifestNotFound } from '../errors';
import { Version } from './version';
import { Source, parseSrc } from './source';
// TODO #features
// import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';
import { manifestOk, manifestInvalid } from '../errors';

export {
  Version,
  Source,
  // TODO #features
  // Feature,
  Dependency,
  Reference,
  Target
};

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 *
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 *
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-properties = { version = "1.0.0" }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 *
 * # [references.Scripting]
 * # version = "1.0"
 * # guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 *
 * [targets]
 * xlsm = "targets/xlsm"
 *
 * [targets.xlam]
 * name = "custom-name"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  type: ManifestType;
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  dir: string;

  // TODO #features
  // features: Feature[];
  // defaultFeatures: string[];
}

export type ManifestType = 'package' | 'project';

const EXAMPLE = `Example vba-block.toml for a package (e.g. library to be shared):

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]

Example vba-block.toml for a project (e.g. workbook, document, etc.):

  [project]
  name = "my-project"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  manifestOk(
    value && (value.package || value.project),
    `[package] or [project] is required, with name, version, and authors specified. \n\n${EXAMPLE}`
  );

  let type: ManifestType, name, version, authors, publish;
  if (value.project) {
    type = 'project';
    name = value.project.name;
    version = value.project.version || '0.0.0';
    authors = value.project.authors || [];
    publish = false;

    manifestOk(name, `[project] name is a required field. \n\n${EXAMPLE}`);
  } else {
    type = 'package';
    name = value.package.name;
    version = value.package.version;
    authors = value.package.authors;
    publish = value.package.publish || false;

    manifestOk(name, `[package] name is a required field. \n\n${EXAMPLE}`);
    manifestOk(
      version,
      `[package] version is a required field. \n\n${EXAMPLE}`
    );
    manifestOk(
      authors,
      `[package] authors is a required field. \n\n${EXAMPLE}`
    );
  }

  const src = parseSrc(value.src || {}, dir);
  const dependencies = parseDependencies(value.dependencies || {}, dir);
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || {}, name, dir);

  // TODO #features
  // const { features, defaultFeatures } = parseFeatures(value.features || {});

  return {
    type,
    name,
    version,
    metadata: { authors, publish },
    src,
    dependencies,
    references,
    targets,
    dir

    // TODO #features
    // features,
    // defaultFeatures,
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');

  if (!(await pathExists(file))) {
    throw manifestNotFound(dir);
  }

  const raw = await readFile(file);

  let parsed;
  try {
    parsed = parseToml(raw.toString());
  } catch (err) {
    const message = `Syntax Error: ${file} (${err.line}:${err.column})\n\n${
      err.message
    }`;
    throw manifestInvalid(message);
  }

  const manifest = parseManifest(parsed, normalize(dir));

  return manifest;
}

export async function writeManifest(manifest: Manifest, dir: string) {
  const { type, name, version, metadata } = manifest;
  const value: any = {
    [type]: { name, version, ...metadata }
  };

  value.src = {};
  manifest.src.forEach(source => {
    let { path, optional } = source;
    path = relative(manifest.dir, path);
    value.src[source.name] = optional ? { path, optional } : path;
  });

  value.targets = {};
  manifest.targets.forEach(target => {
    let { name, type, path } = target;
    path = relative(manifest.dir, path);
    value.targets[type] = name !== manifest.name ? { name, path } : path;
  });

  // TODO dependencies and references

  const toml = convertToToml(value);
  await writeFile(join(dir, 'vba-block.toml'), toml);
}
