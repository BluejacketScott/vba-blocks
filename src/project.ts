import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
