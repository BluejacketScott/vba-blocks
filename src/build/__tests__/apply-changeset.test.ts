import { complex, standardChangesExport } from '../../../tests/__fixtures__';
import { reset, setup } from '../../../tests/__helpers__/project';
import { normalizeManifest } from '../../__helpers__/manifest';
import applyChangeset from '../apply-changeset';
import compareBuildGraphs from '../compare-build-graphs';
import loadFromExport from '../load-from-export';
import loadFromProject from '../load-from-project';

afterEach(reset);

test('should apply changeset for project', async () => {
  jest.spyOn(console, 'log').mockImplementation(() => {});

  const { project, dependencies } = await setup(complex, { silent: false });
  const before = await loadFromProject(project, dependencies);
  const after = await loadFromExport(standardChangesExport);
  const changeset = compareBuildGraphs(before, after);

  await applyChangeset(project, changeset);

  expect(normalizeManifest(project.manifest)).toMatchSnapshot();
});
