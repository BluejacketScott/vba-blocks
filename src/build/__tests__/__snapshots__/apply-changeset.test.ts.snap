// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should apply changeset for project 1`] = `Array []`;

exports[`should apply changeset for project 2`] = `
Array [
  Array [
    "projects/complex/src/ThisWorkbook.cls",
    "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"ThisWorkbook\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attribute VB_Exposed = True
' (changed)
",
  ],
  Array [
    "projects/complex/src/Added.bas",
    "Attribute VB_Name = \\"Added\\"
' (added)
",
  ],
  Array [
    "projects/complex/vba-block.toml",
    "[project]
name = \\"complex\\"
version = \\"0.0.0\\"
authors = [\\"Tim Hall\\"]

[src]
ThisWorkbook = \\"src/ThisWorkbook.cls\\"
Sheet1 = \\"src/Sheet1.cls\\"
Sheet2 = \\"src/Sheet2.cls\\"
Sheet3 = \\"src/Sheet3.cls\\"
UserForm1 = \\"src/UserForm1.frm\\"
Class1 = \\"src/Class1.cls\\"
Added = \\"src/Added.bas\\"

[targets]
xlsm = \\"targets/xlsm\\"
",
  ],
]
`;
